File: apps/nate/app/src/main/java/com/xian/nate/util/Images.kt
/*
 * NATE — Lemonade-powered camera translator (Android).
 * Copyright (C) 2026  Clementine Pendragon <clem@pendragon.systems>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * Contact: clem@pendragon.systems (Clementine Pendragon, c/o Xian Project Development)
 */

package com.xian.nate.util

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import androidx.core.content.FileProvider
import java.io.File

object Images {
    /** Largest edge we send to the model — keeps payload and latency reasonable. */
    private const val MAX_EDGE = 1280

    /** Decode a content Uri to a software bitmap, downscaled to [MAX_EDGE]. */
    fun decode(context: Context, uri: Uri): Bitmap? {
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        context.contentResolver.openInputStream(uri)?.use {
            BitmapFactory.decodeStream(it, null, bounds)
        }
        if (bounds.outWidth <= 0) return null

        var sample = 1
        val longest = maxOf(bounds.outWidth, bounds.outHeight)
        while (longest / sample > MAX_EDGE * 2) sample *= 2

        val opts = BitmapFactory.Options().apply {
            inSampleSize = sample
            inPreferredConfig = Bitmap.Config.ARGB_8888
        }
        val decoded = context.contentResolver.openInputStream(uri)?.use {
            BitmapFactory.decodeStream(it, null, opts)
        } ?: return null

        val edge = maxOf(decoded.width, decoded.height)
        if (edge <= MAX_EDGE) return decoded
        val scale = MAX_EDGE.toFloat() / edge
        return Bitmap.createScaledBitmap(
            decoded, (decoded.width * scale).toInt(), (decoded.height * scale).toInt(), true,
        )
    }

    /** A FileProvider Uri in the cache dir for the camera app to write into. */
    fun newCameraUri(context: Context): Uri {
        val dir = File(context.cacheDir, "captures").apply { mkdirs() }
        val file = File(dir, "capture_${System.currentTimeMillis()}.jpg")
        return FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
    }
}

File: apps/mage-client/src/mage/utils/images.py
# MAGE — Gaming HUD for real-time screen translation.
# Copyright (C) 2026  Clementine Pendragon <clem@pendragon.systems>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# Contact: clem@pendragon.systems (Clementine Pendragon, c/o Xian Project Development)

"""Conversions between Qt image buffers and PIL.

Capture backends hand back a QImage; the vision pipeline wants a PIL image.
Going through an encoded intermediate (PNG or JPEG) costs an encode plus a
decode per frame, which is dead weight in continuous modes — these helpers
copy the pixel buffer across directly instead.
"""

from __future__ import annotations

from PIL import Image
from PyQt6.QtGui import QImage

__all__ = ["qimage_to_pil"]


def qimage_to_pil(image: QImage) -> Image.Image:
    """Convert a QImage to a PIL RGB image without an encode round trip."""
    if image.isNull():
        raise ValueError("Cannot convert a null QImage")

    converted = image.convertToFormat(QImage.Format.Format_RGB888)
    width, height = converted.width(), converted.height()

    ptr = converted.constBits()
    ptr.setsize(converted.sizeInBytes())

    # Qt pads each scanline to a 4-byte boundary, so hand the stride to PIL
    # rather than assuming the rows are contiguous.
    return Image.frombuffer(
        "RGB", (width, height), bytes(ptr), "raw", "RGB", converted.bytesPerLine(), 1
    )
